api/analyzedphenotypes.genus.api.inc

```diff
@@ -0,0 +1,194 @@
+  );
+
+  return ($result->rowCount()) ? $result->fetchCol() : 0;
+}
+
+
+///////
+
+
+/**
+ * CONTROL/SWITCHBOARD.
+ * Pass a request to a corresponding function.
+ * @TODO: leave out when api reorganize/update is complete and use functions below.
+ */
+function analyzedphenotypes_genusprop($property = null, $parameter = null) {
+  if (is_null($property)) {
+    // Default.
+    $property = 'default';
+  }
+
+  // Process request.
+  // Default to all projects.
+  $command = array(
+    'default' => array(
+      'func'  => 'ap_get_genus',
+      'param' => null,
+      'opts'  => null,
+    ),
+    'project_genus' => array(
+      'func'  => 'ap_get_projectgenus',
+      'param' => array('value' => (isset($parameter['project_id'])) ? $parameter['project_id'] : $parameter['project_name']),
+      'opts'  => array('is_id' => (isset($parameter['project_id'])) ? TRUE : FALSE),
+    ),
+    'assign_genus' => array(
+      'func'  => 'ap_set_projectgenus',
+      'param' => array('project_id' => $parameter['project_id'], 'genus' => $parameter['genus']),
+      'opts'  => null,
+    ),
+    'sysvar_genus_set' => array(
+      'func'  => 'ap_get_activegenus',
+      'param' => null,
+      'opts'  => null,
+    ),
+  );
+
+  // Function name, parameter (project name key or name), additional options.
+  return call_user_func($command[$property]['func'], $command[$property]['param'], $command[$property]['opts']);
+}
+
+
+///////
+
+
+/**
+ * FUNCTION:
+ * Assign a genus to an experiment/project in chado.projectprop table.
+ * Dependencies: ap_get_variablenames() - systemvars.api.
+ *
+ * @param $identifiers
+ *   An array with the following keys:
+ *   - project_id : project id number, to which the genus will be linked to.
+ *   - genus : genus.
+ */
+function ap_set_projectgenus($identifiers) {
+  // Fetch the configuration for the term Genus from terms system variable set.
+  $sysvar = ap_get_variablenames(
+    array('variablename' => 'genus'),
+    array('set' => 'terms')
+  );
+
+  $sysvar_genus = variable_get($sysvar);
+
+  // Before relating genus to a project, test for any previous correlations.
+  $is_rel = chado_query(
+    "SELECT projectprop_id FROM {projectprop} WHERE value = :genus AND project_id = :project_id AND type_id = :sysvar_genus",
+    array(
+      ':genus'   => $identifiers['genus'],
+      ':project_id' => $identifiers['project_id'],
+      ':sysvar_genus' => $sysvar_genus,
+    )
+  )
+  ->fetchCol();
+
+  if (!$is_rel) {
+    // Is clear, no past relations made.
+    chado_insert_record('projectprop', array(
+      'project_id' => $identifiers['project_id'],
+      'type_id' => $sysvar_genus,
+      'value' => $identifiers['genus']
+    ));
+  }
+}
+
+/**
+ * FUNCTION:
+ * Fetch genus assigned to an experiment/project in chado.projectprop table.
+ * Dependencies: ap_get_variablenames() - systemvars.api.
+ *
+ * @param $identifiers
+ *   An array, with the following keys:
+ *     - value : project id number or project name.
```

Comment:
This is better done by allowing the key to be either project_id or name depending upon what is passed in. Then you don't need the `is_id` option below because you can tell based on the key. It also makes the function call more readable IMO.

Reply:
like I originally coded it :)